### PR TITLE
feat(discover): Discover prototype - event search

### PR DESCRIPTION
**Pull request by lynnagara**
This is a prototype of "discover" event search functionality, for internal users only.

It allows for very basic searching and filtering of event data across an organization.
It exposes the Snuba API via a new OrganizationDiscover endpoint and introduces a new view to the UI at organization/:orgId/discover/.

This feature is still in a rough state, and is hence feature gated for internal users only.

### CODE REVIEW

File: src/sentry/utils/snuba.py [hunk 1]
```diff
@@ -67,10 +67,9 @@ def query(start, end, groupby, conditions=None, filter_keys=None,
     groupby = groupby or []
     conditions = conditions or []
     having = having or []
-    aggregations = aggregations or [['count()', '', 'aggregate']]
+    aggregations = aggregations or []
```
Comment (lynnagara) [inline src/sentry/utils/snuba.py]: Need to check if anything depended on this default aggregation

**Comment by lynnagara** [comment]
@alex-hofsteede Any advice to deal with this ConnectionError?

File: src/sentry/api/endpoints/organization_discover.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+from sentry.utils import snuba
+
+from django.core.exceptions import ValidationError
+
+
+class OrganizationDiscoverEndpoint(OrganizationEndpoint):
+    permission_classes = (OrganizationDiscoverPermission, )
+
+    def do_query(self, start, end, selected_columns=[], conditions=[], having=[], filters={},
+                 aggregations=[], orderby='-last_seen', limit=None, groupby=[], rollup=None):
```
Comment (macqueen) [inline src/sentry/api/endpoints/organization_discover.py]: you don't want to have mutable data structures as the default args in python, see: http://docs.python-guide.org/en/latest/writing/gotchas/#mutable-default-arguments

File: src/sentry/api/endpoints/organization_discover.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        return snuba_results
+
+    def post(self, request, organization):
+        data = request.DATA
+
+        filters = {
+            'project_id': data['projects']
+        } if 'projects' in data else None
+
+        selected_columns = data['fields'] if 'fields' in data else None
```
Comment (macqueen) [inline src/sentry/api/endpoints/organization_discover.py]: i think you can just shorten this to `selected_columns = data.get('fields')` (and same for all of the ones below)

File: src/sentry/api/endpoints/organization_discover.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+        orderby = data['orderby'] if 'orderby' in data else None
+
+        conditions = data['conditions'] if 'conditions' in data else None
+
+        limit = data['limit'] if 'limit' in data else 1000
```
Comment (macqueen) [inline src/sentry/api/endpoints/organization_discover.py]: here you can do`limit = data.get('limit', 1000)`

File: src/sentry/api/endpoints/organization_discover.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        aggregations = data['aggregations'] if 'aggregations' in data else None
+
+        groupby = data['groupby'] if 'groupby' in data else None
+
+        rollup = data['rollup'] if 'rollup' in data else None
+
+        org_projects = [org.id for org in list(Project.objects.filter(
+            organization=organization,
+            status=ProjectStatus.VISIBLE,
+        ))]
```
Comment (macqueen) [inline src/sentry/api/endpoints/organization_discover.py]: what are you trying to do here? you're querying projects but then using `org` as the variable name

File: src/sentry/api/endpoints/organization_discover.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+        if (not isinstance(limit, int) or limit < 0 or limit > 1000):
+            raise ValidationError('Invalid limit parameter')
```
Comment (macqueen) [inline src/sentry/api/endpoints/organization_discover.py]: what exactly does this do? I _think_ it would wind up causing a 500 error, when i think what you likely want is to do something like `return Response({'detail': 'Invalid limit parameter'}, status=400)` 

My understanding is that you would only want to raise this from a form or model validation: https://docs.djangoproject.com/en/1.7/ref/exceptions/#validationerror

File: src/sentry/api/endpoints/organization_discover.py (hunk 5 above)
Comment (macqueen) [reply src/sentry/api/endpoints/organization_discover.py]: ok, based on your usage below, seems like you are trying to query projects. i think you should filter these based on what the requesting user has access to

Comment (lynnagara) [reply src/sentry/api/endpoints/organization_discover.py]: Oops yeah, i think we should be strict and return a validation error rather than attempt to coerce this value if a user requests a project they can't access though

**Comment by alex-hofsteede** [comment]
@lynnagara sorry I'm missing some context I think. what ConnectionError?

**Comment by lynnagara** [comment]
@alex-hofsteede My bad, if you click on the failed travis test, you'll see that the test tries to connect to localhost:1218 and fails

**Review by macqueen — APPROVED** [review]
left some comments, but otherwise 👍

File: src/sentry/api/endpoints/organization_discover.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+class OrganizationDiscoverPermission(OrganizationPermission):
+    scope_map = {
+        'POST': ['org:read', 'project:read']
+    }
+
+
+class OrganizationDiscoverEndpoint(OrganizationMemberEndpoint):
+    permission_classes = (OrganizationDiscoverPermission, )
+
+    def do_query(self, start, end, selected_columns=None, conditions=None, having=None, filters=None,
```
Comment (alex-hofsteede) [inline src/sentry/api/endpoints/organization_discover.py]: given this is just a passthrough to raw_query, maybe its better to just use `**kwargs` here instead of enumerating them and redoing all the defaults.

File: src/sentry/utils/snuba.py [hunk 8]
```diff
@@ -43,9 +43,9 @@ def timer(name, prefix='snuba.client'):
 )
 
 
-def query(start, end, groupby, conditions=None, filter_keys=None,
-          aggregations=None, rollup=None, arrayjoin=None, limit=None, orderby=None,
-          having=None, referrer=None, is_grouprelease=False):
+def raw_query(start, end, groupby, conditions=None, filter_keys=None,
```
[the remark this answers is not in the document]
Comment (alex-hofsteede) [reply src/sentry/utils/snuba.py]: looking now

File: src/sentry/utils/snuba.py [hunk 9]
```diff
@@ -99,8 +99,11 @@ def query(start, end, groupby, conditions=None, filter_keys=None,
     # we need to fetch the issue definitions (issue -> fingerprint hashes)
     aggregate_cols = [a[1] for a in aggregations]
     condition_cols = [c[0] for c in flat_conditions(conditions)]
-    all_cols = groupby + aggregate_cols + condition_cols
+    all_cols = groupby + aggregate_cols + condition_cols + selected_columns
     get_issues = 'issue' in all_cols
+
+    selected_columns = [col for col in selected_columns if col != 'issue']
```
Comment (alex-hofsteede) [inline src/sentry/utils/snuba.py]: I don't think I quite get why 'issue' is removed here

File: src/sentry/utils/snuba.py [hunk 10]
```diff
@@ -140,10 +144,30 @@ def query(start, end, groupby, conditions=None, filter_keys=None,
...
+    return body
+
+
+def query(start, end, groupby, conditions=None, filter_keys=None,
+          aggregations=None, rollup=None, arrayjoin=None, limit=None, orderby=None,
+          having=None, referrer=None, is_grouprelease=False, selected_columns=None):
 
+    groupby = groupby or []
+    conditions = conditions or []
+    having = having or []
```
Comment (alex-hofsteede) [inline src/sentry/utils/snuba.py]: some of these like `having` which aren't used anywhere in this function except to be passed down to `raw_query` don't need to be modified here.

File: src/sentry/utils/snuba.py (hunk 9 above)
Comment (lynnagara) [reply src/sentry/utils/snuba.py]: oops, thanks. i thought this wasn't supported initially

Comment (lynnagara) [reply src/sentry/utils/snuba.py]: actually... is it?
